lnst/Recipes/ENRT/ConfigMixins/FirewallMixin.py

```diff
@@ -0,0 +1,197 @@
+
+        for host, ruleset in config.firewall_rulesets.items():
+            desc.append("Firewall: ruleset with %d lines on host %s" % (len(ruleset.split("\n")), host))
+
+        return desc
+
+    def remove_sub_configuration(self, config):
+        for host, ruleset in self.stored_rulesets.items():
+            # XXX: this may spam debug log, but counter values may be worth logging
+            logging.debug("Dropping Firewall ruleset on host %s:\n%s\n" % (host, self._dump_ruleset(host, config)))
+            name = self._push_ruleset(host, ruleset)
+            self._apply_ruleset(host, name)
+            host.run('rm -f "%s"' % name)
+
+        return super().remove_sub_configuration(config)
+
+class BulkFirewallMixin(FirewallMixin):
+    """
+    Extend FirewallMixin by the ability to create hosts' firewall configs using
+    a generator property. The generator may yield multiple times to test
+    different ruleset sizes.
+    """
+
+    @property
+    def bulk_firewall_rules_generator(self):
+        """
+        A generator yielding { host: ruleset, ... } to apply for a test run.
+        Each yield will turn into a new sub configuration and thus be tested
+        separately.
+        """
+        raise NotImplementedError("Method bulk_firewall_rules_generator must be defined by a child class.")
+
+    def generate_sub_configurations(self, config):
+        for ruleset in self.bulk_firewall_rules_generator:
+            new_config = copy.copy(config)
+            new_config.firewall_rulesets = ruleset
+            yield new_config
+
+class NftablesMixin(FirewallMixin):
+    """
+    An nftables backend for FirewallMixin.
+    """
+
+    def _apply_ruleset(self, host, path):
+        host.run('nft flush ruleset')
+        host.run('nft -f "%s"' % path)
+
+    def _dump_ruleset(self, host, config):
+        return host.run('nft list ruleset').stdout
+
+class Ip46tablesMixin(FirewallMixin):
+    """
+    A common base class for all the iptables-like FirewallMixin backends.
+    Do not inherit directly, use one of the *tablesMixin classes instead.
+    """
+
+    @property
+    def iptables_command(self):
+        raise NotImplementedError("Property iptables_command must be defined by a child class.")
+
+    def _apply_ruleset(self, host, path):
+        host.run('%s-restore "%s"' % (self.iptables_command, path))
+
+    def _extract_tables_n_chains(self, ruleset):
+        out = {}
+        curtable = None
+        for line in ruleset.split('\n'):
+            if line[0] == '*':
+                out[line] = []
+            elif line[0] == ':':
+                words = line.split(' ')
+                if not words[1] == '-': # ignore user-defined chains
+                    out[curtable].append(words[0][1:])
+        return out
+
+    def _append_missing_parts(self, dst_ruleset, src_ruleset):
+        for table, chains in self._extract_tables_n_chains(src_ruleset).items():
+            tline = '*' + table + '\n'
```

Comment:
the `table` here is coming from https://github.com/LNST-project/lnst/pull/364/files#diff-c1ca3132235b422fab430b7645942020bf1635e314f37db433a3ca03575f4e2dR129 which i think means that it fits the format of `*....` with the `\n` cut off

but here you assign `tline = '*' + "*...."` so you'll end up with double `*` characters is that correct?